src/main/flight/imu.c

```diff
@@ -229,238 +188,218 @@ static void imuCalculateAcceleration(timeDelta_t deltaT)
-        const float a = acc.accADC[axis];
-        accMagnitude += a * a;
-    }
-
-    accMagnitude = accMagnitude * 100 / (sq((int32_t)acc.dev.acc_1G));
-
-    // Accept accel readings only in range 0.90g - 1.10g
-    return (81 < accMagnitude) && (accMagnitude < 121);
-}
-
 static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
 {
     static timeUs_t previousIMUUpdateTime;
-    float rawYawError = 0;
-    bool useAcc = false;
-    bool useMag = false;
-    bool useYaw = false;
-
     const timeDelta_t deltaT = currentTimeUs - previousIMUUpdateTime;
     previousIMUUpdateTime = currentTimeUs;
 
-    if (imuIsAccelerometerHealthy()) {
-        useAcc = true;
-    }
-
-#ifdef USE_MAG
-    if (sensors(SENSOR_MAG) && compassIsHealthy()) {
-        useMag = true;
-    }
-#endif
-#if defined(USE_GPS)
-    else if (STATE(FIXED_WING) && sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat >= 5 && gpsSol.groundSpeed >= 300) {
-        // In case of a fixed-wing aircraft we can use GPS course over ground to correct heading
-        rawYawError = DECIDEGREES_TO_RADIANS(attitude.values.yaw - gpsSol.groundCourse);
-        useYaw = true;
-    }
-#endif
-
 #if defined(SIMULATOR_BUILD) && defined(SKIP_IMU_CALC)
     UNUSED(imuMahonyAHRSupdate);
-    UNUSED(useAcc);
-    UNUSED(useMag);
-    UNUSED(useYaw);
-    UNUSED(rawYawError);
 #else
 
 #if defined(SIMULATOR_BUILD) && defined(SIMULATOR_IMU_SYNC)
 //  printf("[imu]deltaT = %u, imuDeltaT = %u, currentTimeUs = %u, micros64_real = %lu\n", deltaT, imuDeltaT, currentTimeUs, micros64_real());
     deltaT = imuDeltaT;
 #endif
-    float gyroAverage[XYZ_AXIS_COUNT];
-    gyroGetAccumulationAverage(gyroAverage);
-    float accAverage[XYZ_AXIS_COUNT];
-    if (!accGetAccumulationAverage(accAverage)) {
-        useAcc = false;
+    // get sensor data
+#ifndef USE_GYRO_IMUF9001
+    quaternion vError = VECTOR_INITIALIZE;
+    quaternion vGyroAverage;
+    quaternion vAccAverage;
+    gyroGetAverage(&vGyroAverage);
+    accGetAverage(&vAccAverage);
+    DEBUG_SET(DEBUG_IMU, DEBUG_IMU_VACCMODULUS, lrintf((quaternionModulus(&vAccAverage)/ acc.dev.acc_1G) * 1000));
+    if (accIsHealthy(&vAccAverage)) {
+         applyAccError(&vAccAverage, &vError);
     }
-    imuMahonyAHRSupdate(deltaT * 1e-6f,
-                        DEGREES_TO_RADIANS(gyroAverage[X]), DEGREES_TO_RADIANS(gyroAverage[Y]), DEGREES_TO_RADIANS(gyroAverage[Z]),
-                        useAcc, accAverage[X], accAverage[Y], accAverage[Z],
-                        useMag, mag.magADC[X], mag.magADC[Y], mag.magADC[Z],
-                        useYaw, rawYawError);
-
+    applySensorCorrection(&vError);
+    imuMahonyAHRSupdate(deltaT * 1e-6f, &vGyroAverage, &vError);
+#else
+    UNUSED(deltaT);
+    UNUSED(applyAccError);
+    UNUSED(imuMahonyAHRSupdate);
+    qAttitude.w = imufQuat.w;
+    qAttitude.x = imufQuat.x;
+    qAttitude.y = imufQuat.y;
+    qAttitude.z = imufQuat.z;
+    applySensorCorrection(&qAttitude);
```

Comment:
@orneryd 
@kidBrazil 

`applySensorCorrection(&qAttitude);`

I think this is wrong,
it has to be a errorVector in quaternion starage class, not a real quaternion !